### ISSUE TITLE
Article Viewer displays raw wikicode for "Health effects of electronic cigarettes"

### ISSUE THREAD

**Issue by ragesoss** [WikiEducationFoundation/WikiEduDashboard#5957]
Visit here and wait for the authorship highlighting to load: https://dashboard.wikiedu.org/courses/UCSF/Foundations_II_(Summer_2024)/articles/edited?showArticle=52260526

Once it loads, the rendered article is replaced by highlighted wikicode:

![Screenshot from 2024-09-10 12-12-23](https://github.com/user-attachments/assets/672d1c74-2999-4d26-a218-9c42bf13e332)

### Additional context

The ArticleViewer initially loads the parsed version of the current article, and requests the authorship data from the wikiwho server. Once received, the wikiwho data (which is annotated wikicode) gets processed by Dashboard code to add CSS classes on a per-author basis, it is sent to mediawiki to parse. No explicit errors are occurring in this example, either in the JS console or in network requests, but the call to the mediawiki API `parse` action is returning unparsed wikicode. One possible explanation is that the Dashboard code that operates on the wikiwho data is mishandling some particular aspect of this page's wikicode, resulting in a version that can't be parsed properly by mediawiki.

**Comment by ragesoss** [comment]
Here's another example: https://dashboard.wikiedu.org/courses/UCSD/Introduction_to_Policy_Analysis_-_Summer_Session24_(Summer)/articles/edited?showArticle=1007667

**Comment by empty-codes** [comment]
Hello @ragesoss, I would like to try working on this!

**Comment by ragesoss** [comment]
@empty-codes go for it. This one may be a challenge, as I'm not sure which codebase is ultimately responsible for the error. There's almost certainly something about the wikicode for these example articles that is triggering the bug, so just knowing precisely what triggers it would be helpful.

**Comment by Abishekcs** [comment]
Hi @ragesoss and @empty-codes,

I hope you're doing well! I wanted to share an observation I made while looking into this bug earlier today.

It seems that for the articles where this issue occurs most frequently, `WikiWho` is rendering the `sup tag` incorrectly, showing:

a) Opening tag: `&lt;ref&gt;  (where it should be <sup>)`
b) Closing tag:  `&lt;/ref&gt;  (where it should be </sup>)`
c) Self-closing tag:  `&lt;ref /&gt;  (where it should be <sup />)`

For example in the actual HTML output, this  appear  as: `<ref />` .

Interestingly, in articles where this bug does not occur, `WikiWho outputs the <sup> tag correctly as <sup>`.

Additionally, I've noticed that the bug tends to happen in articles where `contributors have also added the citations`.
However, I did encounter at least one article where a `citation didn't cause this issue`, though I can't recall which article it was.

Please consider these as initial thoughts— as I’ve been trying to understand how WikiWho algorithm works. Hope my explanation was clear

Good luck with solving the bug @empty-codes!

**Comment by empty-codes** [comment]
@Abishekcs Thank you so much for your insights; they were really helpful for getting started.  
I still have not solved the bug, but these are my findings so far @ragesoss : (sorry it's a bit long)

Firstly, I found that there are actually **two different APIs** involved:

1. The [WikiWhoAPI](https://wikiwho-api.wmcloud.org/en/api/v1.0.0-beta/#/) which is designed to parse historical revisions of Wikipedia articles, providing detailed provenance of each token (word) in terms of who added, removed, or reintroduced it across different revisions.

2. The [WhoColorAPI](https://wikiwho-api.wmcloud.org/en/whocolor/v1.0.0-beta/#/) which is built on top of the WikiWho API and allows for the visualization of authorship data by color-coding tokens in the text based on their original authors. Wiki Edu Foundation employs this to show authorship data on its dashboard for students.

For this issue, the **WhoColorAPI** is the one we're concerned with.

### Flow:

1. Initially, the `ArticleViewer` component loads the parsed version of the article:

   ```javascript
   <div id="article-scrollbox-id" className="article-scrollbox">
     {
       fetched ? <ParsedArticle highlightedHtml={highlightedHtml} whocolorHtml={whoColorHtml} parsedArticle={parsedArticle} /> : <Loading />
     }
   </div>
   ```

   The `ParsedArticle` component is defined in `ParsedArticle.jsx`: 

   ```javascript
   export const ParsedArticle = ({ highlightedHtml, whocolorHtml, parsedArticle }) => {
     let articleHTML = highlightedHtml || whocolorHtml || parsedArticle;
   ```

   The `ParsedArticle` component accepts `highlightedHtml`, `whocolorHtml`, and `parsedArticle` as props and displays one of them based on what is available.

2. It then fetches authorship data from the WikiWho server.

3. Once the authorship data is available, it replaces the initially rendered parsed article with the highlighted HTML (from `whoColorHtml`).

### Here's the difference between the three props:

- **`parsedArticle`**: This is the basic version of the article fetched by the `fetchParsedArticle` method that is initially rendered. This is just the plain article HTML without any authorship highlighting obtained from the MediaWiki API call using the `parsedArticleURL(lastRevisionId)` method in the `URLBuilder`, which returns a URL of this format:

   ```javascript
   `${base}/w/api.php?action=parse&oldid=${lastRevisionId}&disableeditsection=true&format=json`;
   ```

- **`whoColorHtml`**: This is the raw HTML returned directly by the WhoColor API. It includes the token-level spans that identify which editor added or modified specific parts of the text. This is obtained from `articleviewer.jsx` by calling the `fetchWhocolorHtml()` function, which further calls the `__wikiwhoColorURLTimedRequestPromise(timeout, lastRevisionId)` function, which uses the `wikiwhoColorURL` URLBuilder method to get the URL of the format: 

   ```javascript
   const url = `${WIKIWHO_DOMAIN}/${language}/whocolor/v1.0.0-beta/${encodeURIComponent(title)}/${revisionId}/`;
   ```

- **`highlightedHtml`**: This is the processed version of `whoColorHtml`, where additional formatting and styling are applied to correctly display the authorship data (e.g., wrapping spans for tokens with additional attributes for styling). It is populated by the `highlightAuthors` function, which uses the `whoColorHtml` state:

   ```javascript
   // This takes the extended_html from the whoColor API, and replaces the span
   // annotations with ones that are more convenient to style in React.
   // The matching and replacing of spans is tightly coupled to the span format
   // provided by the whoColor API: https://github.com/wikimedia/wikiwho_api
   const highlightAuthors = () => {
     let html = whoColorHtml;
      // Replace each editor span for this user with one that includes their
      // username and color class.
      const prevHtml = html;
      const colorClass = colors[i];
      const styledAuthorSpan = `<span title="${user.name}" class="editor-token token-editor-${user.userid} ${colorClass}`;
      const authorSpanMatcher = new RegExp(`<span class="editor-token token-editor-${user.userid}`, 'g');
      html = html.replace(authorSpanMatcher, styledAuthorSpan);
     // more logic and logic and logic 
     setHighlightedHtml(html); // highlightedHtml state variable populated here
     setPendingRequest(false);
   };
   ```

@Abishekcs identified that the problem seems to occur because WikiWho is incorrectly outputting `<ref>` tags instead of `<sup>` tags. This misrepresentation leads to HTML errors when the browser attempts to render the content since `<ref>` is not a recognized standard HTML tag.

Note that the MediaWiki parse action returns a `parse.text` property that correctly contains all the `<sup>` tags for both the affected and unaffected articles, which is why the page renders fine initially as the `parsedArticle` prop/state variable. So the problem is likely not from the MediaWiki API.

The fact that some articles display correctly while others do not suggests that there may be inconsistencies in how the WhoColor API processes certain revisions of articles. 

**So the questions are:**

1. Is it the `whoColorHtml` that replaces parts of the `parsedArticle` that are meant to be `<sup>` with `<ref>`, or is it the `highlightedHtml` that replaces the `<sup>` with `<ref>`? 

2. What exactly is it about the affected articles that is triggering the bug?

If it is the `whoColorHtml` replacing it, it is likely an issue with the WhoColor API; if it is the `highlightedHtml`, it is likely an issue with the `highlightAuthors` function logic in `ArticleViewer.jsx`.

Additionally, @Abishekcs also noticed that the bug tends to happen in articles where contributors have also added citations.

In this specific article: [UCSF Foundations II](https://dashboard.wikiedu.org/courses/UCSF/Foundations_II_(Summer_2024)/articles/edited?showArticle=52260526), in the parse action response, there were parse warnings:

```javascript
parsewarnings[ 
  "Script warning: <span style=\"color:#3a3\">One or more <code style=\"color: inherit; background: inherit; border: none; padding: inherit;\">&#123;{[[Template:cite journal|cite journal]]}}</code> templates have maintenance messages</span>; messages may be hidden ([[Help:CS1_errors#Controlling_error_message_display|help]])."
]
```

Since the bug appears more frequently in articles with numerous citations, it’s possible that these templates are not being parsed correctly. However, these parse warnings are inconsistent (and probably irrelevant) because they were present in another article that does not have this bug and also absent in another article that has this bug.

Keeping all these in mind, I will continue further investigation. Hopefully I can  pinpoint a cause soon😅

**Comment by empty-codes** [comment]
> @empty-codes, I believe it's the whoColorHtml. I'm mentioning this because 😅 I reviewed the raw HTML output for whoColorHtml, and here's a small screenshot below. However, it might be a good idea to cross-check just to be sure.

That answers the question, thank you! I'll also crosscheck from my end.

**Comment by empty-codes** [comment]
I was stuck trying to find a lead for a while 😅 but I finally got somewhere (I think).

For the 1st [affected article:](https://dashboard.wikiedu.org/courses/UCSD/Introduction_to_Policy_Analysis_-_Summer_Session24_(Summer)/articles/edited?showArticle=1007667) the whocolorAPI [ GET Request Method](https://wikiwho-api.wmcloud.org/en/whocolor/v1.0.0-beta/Hispanic%20and%20Latino%20Americans/0/?origin=*) returns this at the end of the extended html response as a comment:

![image](https://github.com/user-attachments/assets/c1b77630-9a61-47c6-8473-757b335d9e9f)

For the 2nd [affected article:](https://dashboard.wikiedu.org/courses/UCSF/Foundations_II_(Summer_2024)/articles/edited?showArticle=52260526)  the whocolorAPI [GET Request Method](https://wikiwho-api.wmcloud.org/en/whocolor/v1.0.0-beta/Health%20effects%20of%20electronic%20cigarettes/0/?origin=*) also returns a report

![image](https://github.com/user-attachments/assets/5ebcca41-75bb-49fe-9459-aef728cba4fd)

For unaffected articles however (I checked about 30~), they tend to have a report with format similar to this:

![image](https://github.com/user-attachments/assets/0259a7aa-c747-47f6-a600-655d85522c8d)

or even this:

![image](https://github.com/user-attachments/assets/6649d430-7c6b-423f-a1fd-d9df1d49671f)


It is seen that for affected articles, `Post‐expand include size`, `Template argument size`, `Expensive parser function count` and so on, are all `0`, `Highest expansion depth` is `1`, `Transclusion expansion time repor`t takes `0ms with 1 call and no templates showing`  so this issue likely stems from the templates which are missing / contain errors / are not invoked correctly / are not expanded.

At this point, I would like to ask for further guidance @Abishekcs @ragesoss , what steps should I take from here please?
Also, i found this file that seems to contain the parsing logic:  [markuppreparser.inc.php](https://github.com/WikiEducationFoundation/whoCOLOR/blob/master/server/markuppreparser.inc.php), please is this still in use and would it be relevant to this issue?

Thank you in advance!

**Comment by ragesoss** [comment]
That's interesting. The template expansion seems like a good clue, it's not obvious to me whether an expansion limit is involved, or whether it's being misparsed for some other reason. The unparsed `ref` tags seem likely to be relevant.

I can't tell whether that whoCOLOR repository is indirectly used for this. The main repo for the wikiwho-api servers is https://github.com/wikimedia/wikiwho_api

**Comment by empty-codes** [comment]
Noted! I will update you on any new findings @ragesoss

**Comment by empty-codes** [comment]
@ragesoss Sorry for the lack of updates; I've been recovering from a minor cold but I'm back to working on this. 

I’ve been focusing on identifying what triggers the bug on the affected pages, but I haven’t made much progress yet. The main clue is that templates on these pages are not expanding correctly, which seems to lead to `<sup>` tags being misinterpreted as `<ref>` tags.

In examining the Hispanic article, I found that in the working Wikipedia edit page, the template is:

```
{{Redirect-multi|2|Latinas|Latinos|other uses|Latina (disambiguation){{!}}Latina|and|Latino (disambiguation){{!}}Latino}}
```

However, in the WhoColor HTML output, it appears as:

```
{{Redirect-multi|2|Latinas|Latinos|other uses|Latina (disambiguation){{!}}<span class="editor-token token-editor-1152308" id="token-53">Latina</span><span class="editor-token token-editor-22831189" id="token-54">|</span>...}}
```

The presence of `editor-token` elements in the template definition seems to indicate that they are being improperly injected into the template content, likely interfering with normal rendering.

My current hypothesis is that, as seen in the [`wikiwho_api\api\tasks.py`](https://github.com/wikimedia/wikiwho_api/blob/main/api/tasks.py) file, the tasks are constrained by soft time limits and cache timeouts, which could lead to incomplete template expansion for articles with many citations or heavy wikitext usage. If tokenization kicks in before the full expansion, it may cause improper injection of elements like editor-token.

Currently, I have set up the WikiWho API locally to test this hypothesis, I will update on my findings!

**Comment by empty-codes** [comment]
Hello @ragesoss, Here is my current update.

Initially, I attempted to resolve the issue of the parser prematurely closing templates upon encountering `{{!}}`  by modifying the `__parse_wiki_text` method in `parser.py` and introducing a template depth counter (`self.template_depth = 0`) and a template stack (`self.template_stack = []`), along with additional logic to track open templates. 

However, this approach resulted in another bug involving duplicate `{{` and `}}` template tags, prompting me to revert the changes. Instead, I added new markup in `special_markups.py` to specifically address template delimiters like `{{!}}`:

```python
{
    'type': 'single',
    'start_regex': re.compile(r'{{!}}'),
    'end_regex': None,
    'no_spans': True,
    'no_jump': False
},
```

This modification effectively eliminated the unwanted `<span>` tags injected within templates; however, the `<ref>` tag bug persists.

Both the `rev_text` and `wiki_text` values utilize correctly formatted `<ref>` tags. However, the bug manifests when generating extended HTML:
   
   ```python
   parser = WikiMarkupParser(wiki_text, whocolor_data['tokens'])
   parser.generate_extended_wiki_markup()
   extended_html = wp_rev_text_obj.convert_wiki_text_to_html(parser.extended_wiki_text)
   ```

By changing the argument of the `convert_wiki_text_to_html` function to `wiki_text`, the templates were expanded correctly. This suggests a potential issue with the `extended_wiki_text` generated by the parser. 

The following are steps I have taken in investigating the bug: 

1. I utilized the [WikiTemplate UDL tool](https://en.wikipedia.org/wiki/User:Equazcion/WikiTemplate_UDL) to compare the wikitexts (both the `wikitext` and the `parser.extended_wiki_text` of affected and unaffected pages, confirming no significant differences in template formats.

2. I attempted the following actions without success:
   - Adding an additional action=expandtemplates request to the handler and utils .py to make a call to the mediawiki expand templates action first to expand all the templates in the wikitext and then pass the result from that to the action=parse, but it did not resolve the issue.
   - Increasing the request header timeout from 0 to 180 seconds, which also yielded no improvements.
   - Modifying the `default_task_soft_time_limit` in `deployment/celery_config.py` from 120 to 300 seconds, yet no change was observed.
   - Uncommenting redundant markup for `<ref>` tags in the markup file, which had no impact.

3. I also utilized the [Wikipedia Special:ExpandTemplates](https://en.wikipedia.org/wiki/Special:ExpandTemplates) tool with both the wikitext and the expanded wikitext (including injected editor tokens) of a buggy page. The templates in the response expanded correctly with both of the input wikitexts, indicating that the injected editor tokens are not likely to be responsible for the problem.

In conclusion, I cannot pinpoint a cause because:

1. If the cause is attributed to parsing logic, I found no evidence supporting this claim, as the same dashboard logic works on other pages, and templates expand properly using the Special:ExpandTemplates tool.
2. If the cause is attributed to excessive template usage and citations, the same page with the same templates and citations renders correctly on Wikipedia and also the initial parsed article view.
3. If the cause is attributed to timeouts, increasing request timeout and task limits did not yield any improvements.

I am still trying to find any loophole I must have missed but I would greatly appreciate any insights or guidance you can provide to help me move forward. Thank you for your patience!

**Comment by ragesoss** [comment]
@empty-codes thanks! this is really useful documentation of your debugging work. I suggest leaving this one; hopefully we can find the next clue at a later time, but it's a relatively rare bug.

I just checked the second example with the Who Wrote That? tool on Wikipedia, and it also displays this buggy behavior (which makes sense based on your debugging, as it's clearly a problem with the WikiWho processing). So we can be pretty confident now that it's not a bug in our codebase.

![wwt wikicode](https://github.com/user-attachments/assets/2d8a5df0-856d-4f12-acf2-72cc35ffc1e9)

One really useful way to wrap this up would be to open an issue on Phabricator against the Who-Wrote-That project, summarizing what you've learned about the like source of the bug within the WikiWho parser. There are some other issues there already related to pages that don't work as expected, but I don't see any that are clearly the same issue here, and I didn't spot anything along the lines of what you've done here to narrow down the source.

**Comment by empty-codes** [comment]
@ragesoss I've created the Phabricator issue here: https://phabricator.wikimedia.org/T377898

While I wasn't able to completely pinpoint the source of the bug, I learned a lot throughout the process and I'm glad this documentation will be useful. Thanks for your guidance throughout this process! 🙏

**Comment by ragesoss** [comment]
Another example: "Donald Trump 2024 presidential campaign" 

https://dashboard.wikiedu.org/courses/Wayne_State_System/Contemporary_Rhetorical_Criticism_(Fall_2024)/articles/edited?showArticle=707429823
